### ISSUE TITLE
Question on Merge, Stack, Concatenate

### ISSUE THREAD

**Issue by coppolajj** [danielhrisca/asammdf#107]
# Python version
```python
 'python=3.7.0 (default, Jun 28 2018, 08:04:48) [MSC v.1912 64 bit (AMD64)]'
 'os=Windows-7-6.1.7601-SP1'
 'numpy=1.15.4'
 'asammdf=4.3.0'
```
# Code 

  ## MDF version
4.10

  
# Description

What's the difference between merge, stack, concatenate?  Also is there an index that is generated when combining files/mdfs so that data points can be traced back to specific files?

**Comment by danielhrisca** [comment]
Hello @coppolajj,

it seems that the docstrings are not clear enough so I'll have to update them.

_merge_ and _concatenate_ do the same thing: all the files have to have the same internal structure (number of channel groups, number of channels in each channel group) and the channels in the output file will have the samples resulted from the concatenation of the samples from all the input files.

_stack_ takes all channel from the input files and the internal structure does not matter.


Try out this snippet to get a feel of the difference

```python
from asammdf import MDF, Signal
import numpy as np


for i in range(2):
    m = MDF()
    s = Signal(
        samples=np.arange(10) + 80*i,
        timestamps=np.arange(10) + 80*i,
        name='Channel',
    )
    m.append(s)
    m.save('f{}.mf4'.format(i), overwrite=True)

concatenated = MDF.concatenate(['f0.mf4', 'f1.mf4'])
print(len(concatenated.groups))
print(concatenated.get('Channel'))

stacked = MDF.stack(['f0.mf4', 'f1.mf4'])
print(len(stacked.groups))
print(stacked.get('Channel', group=0))
print(stacked.get('Channel', group=1))
```

**Comment by danielhrisca** [comment]
I forgot to answer the second question: there is no way to trace back the samples to the original files they came from. Can you give a context in which you would need to?

**Comment by coppolajj** [comment]
Thanks for clarifying!  This explains precisely what I was seeing in my initial tests (although I couldn't figure out what was going on with stack).

One case where you might want to reference the original file is if you see an anomaly in the data, you may want to go back to the original file and view it in another tool (MDA, vSignalyzer, etc) as a sanity check.

Another case is for calculating the difference between adjacent elements.  The difference between the last element of the first file and the 1st element of the 2nd file is not really relevant (I usually set this to NaN or zero).  But when I use merge/concatenate the original time is lost so I don't know where the second file starts.   

If the original times were left in place then it would be clear where the file boundaries are... maybe this could be done with concatenate???

When exporting the stacked data to a pandas dataframe it seems that the groups are lost, maybe they can be used as keys?

Thanks so much, this is a great tool!

**Comment by danielhrisca** [comment]
Hello @coppolajj ,

I've made a quick implementation to hold a minimal trace to the original file. The info is stored in the channel group comment.

Running this
```python
from asammdf import MDF, Signal
import numpy as np
from pprint import pprint

for i in range(2):
    m = MDF()
    s = Signal(
        samples=np.arange(10) + 80*i,
        timestamps=np.arange(10) + 80*i,
        name='Channel',
    )
    m.append(s)
    m.save('f{}.mf4'.format(i), overwrite=True)

concatenated = MDF.concatenate(['f0.mf4', 'f1.mf4'])
pprint(concatenated.info())

stacked = MDF.stack(['f0.mf4', 'f1.mf4'])
pprint(stacked.info())
```
the output is this
```python
{'group 0': {'channel 0': 'name="time" type=master',
             'channel 1': 'name="Channel" type=value',
             'channels count': 2,
             'comment': '0.000000s to 9.000000s concatenated from channel '
                        'group 0 of "f0.mf4"\n'
                        '80.000000s to 89.000000s concatenated from channel '
                        'group 0 of "f1.mf4"\n',
             'cycles': 20},
 'groups': 1,
 'version': '4.10'}
{'group 0': {'channel 0': 'name="time" type=master',
             'channel 1': 'name="Channel" type=value',
             'channels count': 2,
             'comment': 'stacked from channel group 0 of "f0.mf4"',
             'cycles': 10},
 'group 1': {'channel 0': 'name="time" type=master',
             'channel 1': 'name="Channel" type=value',
             'channels count': 2,
             'comment': 'stacked from channel group 0 of "f1.mf4"',
             'cycles': 10},
 'groups': 2,
 'version': '4.10'}
```

Is this what you needed?

**Comment by coppolajj** [comment]
I think that would work... although I wouldn't mind if there was an additional signal that had the original time stamps stacked on top of each other (or an option to create it).

**Comment by danielhrisca** [comment]
I think that adding new channels would complicate things.

I propose for now to use the improved comment with the latest commit and extract the original samples

```python
from asammdf import MDF, Signal
import numpy as np
from pprint import pprint

for i in range(2):
    m = MDF()
    s = Signal(
        samples=np.arange(10) + 80*i,
        timestamps=np.arange(10) + 80*i,
        name='Channel',
    )
    print(s)
    m.append(s)
    m.save('f{}.mf4'.format(i), overwrite=True)

concatenated = MDF.concatenate(['f0.mf4', 'f1.mf4'])
pprint(concatenated.info())

from_first_file = concatenated.get('Channel').cut(start=0, stop=9)
from_first_file.timestamps = from_first_file.timestamps - 0 + 0
from_second_file = concatenated.get('Channel').cut(start=80)
from_second_file.timestamps = from_second_file.timestamps - 80 + 80

print(from_first_file)
print(from_second_file)

print('='*40)

for i in range(2):
    m = MDF()
    s = Signal(
        samples=np.arange(10) + 80*i,
        timestamps=np.arange(10) + 5*i,
        name='Channel',
    )
    print(s)
    m.append(s)
    m.save('f{}.mf4'.format(i), overwrite=True)

concatenated = MDF.concatenate(['f0.mf4', 'f1.mf4'])
pprint(concatenated.info())

from_first_file = concatenated.get('Channel').cut(start=0, stop=9)
from_first_file.timestamps = from_first_file.timestamps - 0 + 0
from_second_file = concatenated.get('Channel').cut(start=10)
from_second_file.timestamps = from_second_file.timestamps - 10 + 5

print(from_first_file)
print(from_second_file)

```

with the result

```
<Signal Channel:
	samples=[0 1 2 3 4 5 6 7 8 9]
	timestamps=[0 1 2 3 4 5 6 7 8 9]
	unit=""
	conversion=None
	source=None
	comment=""
	mastermeta="None"
	raw=True
	display_name=
	attachment=()>

<Signal Channel:
	samples=[80 81 82 83 84 85 86 87 88 89]
	timestamps=[80 81 82 83 84 85 86 87 88 89]
	unit=""
	conversion=None
	source=None
	comment=""
	mastermeta="None"
	raw=True
	display_name=
	attachment=()>

{'group 0': {'channel 0': 'name="time" type=master',
             'channel 1': 'name="Channel" type=value',
             'channels count': 2,
             'comment': '0.0s to 9.0s concatenated from channel group 0 of '
                        '"f0.mf4" with first time stamp at 0.0s\n'
                        '80.0s to 89.0s concatenated from channel group 0 of '
                        '"f1.mf4" with first time stamp at 80.0s\n',
             'cycles': 20},
 'groups': 1,
 'version': '4.10'}
<Signal Channel:
	samples=[0 1 2 3 4 5 6 7 8 9]
	timestamps=[0. 1. 2. 3. 4. 5. 6. 7. 8. 9.]
	unit=""
	conversion=None
	source=None
	comment=""
	mastermeta="('time', 1)"
	raw=False
	display_name=
	attachment=()>

<Signal Channel:
	samples=[80 81 82 83 84 85 86 87 88 89]
	timestamps=[80. 81. 82. 83. 84. 85. 86. 87. 88. 89.]
	unit=""
	conversion=None
	source=None
	comment=""
	mastermeta="('time', 1)"
	raw=False
	display_name=
	attachment=()>

========================================
<Signal Channel:
	samples=[0 1 2 3 4 5 6 7 8 9]
	timestamps=[0 1 2 3 4 5 6 7 8 9]
	unit=""
	conversion=None
	source=None
	comment=""
	mastermeta="None"
	raw=True
	display_name=
	attachment=()>

<Signal Channel:
	samples=[80 81 82 83 84 85 86 87 88 89]
	timestamps=[ 5  6  7  8  9 10 11 12 13 14]
	unit=""
	conversion=None
	source=None
	comment=""
	mastermeta="None"
	raw=True
	display_name=
	attachment=()>

{'group 0': {'channel 0': 'name="time" type=master',
             'channel 1': 'name="Channel" type=value',
             'channels count': 2,
             'comment': '0.0s to 9.0s concatenated from channel group 0 of '
                        '"f0.mf4" with first time stamp at 0.0s\n'
                        '10.0s to 19.0s concatenated from channel group 0 of '
                        '"f1.mf4" with first time stamp at 5.0s\n',
             'cycles': 20},
 'groups': 1,
 'version': '4.10'}
<Signal Channel:
	samples=[0 1 2 3 4 5 6 7 8 9]
	timestamps=[0. 1. 2. 3. 4. 5. 6. 7. 8. 9.]
	unit=""
	conversion=None
	source=None
	comment=""
	mastermeta="('time', 1)"
	raw=False
	display_name=
	attachment=()>

<Signal Channel:
	samples=[80 81 82 83 84 85 86 87 88 89]
	timestamps=[ 5.  6.  7.  8.  9. 10. 11. 12. 13. 14.]
	unit=""
	conversion=None
	source=None
	comment=""
	mastermeta="('time', 1)"
	raw=False
	display_name=
	attachment=()>
```

**Comment by nos86** [comment]
Hello Daniel,
on this item, I have a doubt.
My use case is the following: I have a measure split in several MDF files. Every file has a timestamp starting from 0.
Assuming that there is no hole between the end of one the start of the next, I'd like to concatenate all files in order to obtain a single measurement file. How can I achieve it using your library? (For the moment, we can assume that all files have the same structure)

Thanks in advance,
Salvo

**Comment by nos86** [comment]
Hi Daniel,
thank you very much for your help and sorry for my late answer.
I agree with your solution; unfortunately I've been involved in another project, and I'm not able to provide you a prompt feedback.

I'll write you in case I found special cases.

Salvo

**Comment by due303** [comment]
Hello Daniel! First of all , thank you for the awesome library, it really helps a lot.
I'm using it to concatenate files from a continuous acquisition and it works perfect. But I would also like to concatenate files from different days and I can't make it work. I have the same data structure (same labels, time resolution, channels, etc) What only differs is the arrange of the signals in each channel, due to my recording software. I always get the same error: "internal structure of file 1 is different; different channels". Is there any work around?
Thank you.
Edwin.

**Comment by danielhrisca** [comment]
Hello @due303 ,

The only way in this case is to manually concatenate the signals and create a new MDF, but the next method will introduce interpolation 

```python
from asammdf import MDF, Signal
import numpy as np

filenames = [f1, f2, f3] # ...
count = len(filenames)
mdfs = sorted([MDF(name) for name in filenames])

channel_names = [c1, c2, c3, c4] # ...

all_signals = [mdf.select(channel_names) for mdf in mdfs]

for mdf in mdfs:
    mdf.close()
    
mdf = MDF()
signals = []

for i, name in enumerate(channel_names):
    sig = Signal(
        samples=np.concatenate(
            [all_signals[index][i].samples for index in range(count)]
        ),
        timestamps=np.concatenate(
            [all_signals[index][i].timestamps for index in range(count)]
        ),
        name=name,
        unit=all_signals[0][i].unit,
    )
        
    signals.append(sig)
    
mdf.append(signals)

mdf.save('concatenated.mf4')

mdf.close()
```

**Comment by due303** [comment]
Hello Daniel, 
Thank you for you fast answer. I have two questions, is there a function to list all channels available in one mdf4? If there is, I couldn't find it. And I'm having trouble with signals that are strings. I found the argument "encoding = str" for the function Signal, which seems to work, but I can't manage to work with the select function, when I set "ignore_value2text_conversions = False" I get the following error: "MdfException: wrong encoding "<class 'str'>" for string signal."

Thanks, Edwin.

**Comment by slicksson** [comment]
Hi!
If I want to merge two mf4 files togheter, and they have a timestamp gap. For example:

timestamp1: [0, 1, 2, 3, 4, 5]
timestamp2: [10, 11, 12, 13, 14, 15]

I want the merged file to have no gaps. I mean that I want timestamp2 start right after the first one: timestamp2 = [6, 7, 8, 9, 10, 11].
Is there any easy way to do that?

I have several files I want to merge and the size of the gaps differ.


Thanks,
J

**Comment by danielhrisca** [comment]
@slicksson use `concatenate` with `sync=False` (https://asammdf.readthedocs.io/en/latest/api.html#asammdf.mdf.MDF.concatenate)

**Comment by slicksson** [comment]
> @slicksson use `concatenate` with `sync=False` (https://asammdf.readthedocs.io/en/latest/api.html#asammdf.mdf.MDF.concatenate)

I tried that but its not solving this issue. 
Try run this and you will see:

----------------------------------------------------------------------------------
from asammdf import MDF, Signal

s = [1,2,3,4,5]
t1 = [0, 1, 2, 3, 4]
t2 = [i + 10 for i in t1]

m1 = MDF()
m2 = MDF()

s1 = Signal(samples = s, timestamps = t1, name='Chanel')
m1.append(s1)

s2 = Signal(samples = s, timestamps = t2, name='Chanel')
m2.append(s2)

conc = MDF.concatenate([m1, m2], sync=False)
conc.save('merged.mf4')

s3 = conc.get('Chanel')
s3.plot()
print(s3.timestamps)
----------------------------------------------------------------------------------

**Comment by danielhrisca** [comment]
Please try the development branch code. You need to use the new "direct_timestamp_continuation" argument

```python
from asammdf import MDF, Signal

s = [1,2,3,4,5]
t1 = [0, 1, 2, 3, 4]
t2 = [i + 10 for i in t1]

m1 = MDF()
m2 = MDF()

s1 = Signal(samples = s, timestamps = t1, name='Chanel')
m1.append(s1)

s2 = Signal(samples = s, timestamps = t2, name='Chanel')
m2.append(s2)

conc = MDF.concatenate([m1, m2], sync=False, direct_timestamp_continuation=True)
conc.save('merged.mf4')

s3 = conc.get('Chanel')
s3.plot()
print(s3.timestamps)
```

**Comment by slicksson** [comment]
It is still the same. The timestamps for s3 = [ 1.  2.  3.  4.  5. 11. 12. 13. 14. 15.].

**Comment by danielhrisca** [comment]
I don't think you have the latest development branch code:

```
[0. 1. 2. 3. 4. 5. 6. 7. 8. 9.]
```

**Comment by slicksson** [comment]
How can I get the latest?

**Comment by danielhrisca** [comment]
`pip install -I --no-deps https://github.com/danielhrisca/asammdf/archive/development.zip`

**Comment by slicksson** [comment]
It's working fine now. Thank you!! :)

**Comment by slicksson** [comment]
New problems. After the installation and _direct_timestamp_continuation=True_ the merging worked when I created new signals and mf4 files as I showed above. But now when I want to merge some other mf4-files, I got this error:


  File "C:\ProgramData\Anaconda3\lib\site-packages\asammdf\mdf.py", line 377, in _transfer_header_data
    self.header.author = other.header.author

AttributeError: 'NoneType' object has no attribute 'author'


It works when I use stack, insted of concatenate. But before installing the updates you sent me @danielhrisca, it worked with concatenate (except the gaps). 
Any ideas?
